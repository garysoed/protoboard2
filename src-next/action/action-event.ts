import {Action} from './action';

export const ACTION_EVENT = 'pb-action';

export class ActionEvent extends Event {
  constructor(
      readonly action: Action<any, any>,
      readonly id: {},
  ) {
    super(ACTION_EVENT, {bubbles: true});
  }
}
