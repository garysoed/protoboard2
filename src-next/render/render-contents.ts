import {Vine} from 'grapevine';
import {RenderSpec} from 'persona';
import {combineLatest, Observable} from 'rxjs';
import {map} from 'rxjs/operators';

import {$getRenderSpec$} from './render-component-spec';


export function renderContents(
    contentIds$: Observable<ReadonlyArray<{}>>,
    vine: Vine,
): Observable<readonly RenderSpec[]> {
  return combineLatest([
    contentIds$,
    $getRenderSpec$.get(vine),
  ])
      .pipe(
          map(([contentIds, getRenderSpec]) => {
            if (!getRenderSpec) {
              throw new Error('getRenderSpec is not initialized');
            }
            return contentIds.map(getRenderSpec);
          }),
      );
}


// return nodes$.pipe(
//   map(renderSpecs => $pipe(
//       renderSpecs,
//       $filterNonNull(),
//       $map(({id, spec}) => {
//         const decorator: Decorator<NodeWithId<Node>> = () => {
//           $setParent.get(vine)(id, containerId);
//           return EMPTY;
//         };

//         const decorators: Array<Decorator<NodeWithId<any>>> = [decorator];
//         if (spec.decorators) {
//           decorators.push(...spec.decorators);
//         }

//         return {...spec, decorators};
//       }),
//       $asArray(),
//   )),
// );