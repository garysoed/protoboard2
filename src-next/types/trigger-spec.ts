export enum TriggerType {
  A = 'a',
  B = 'b',
  C = 'c',
  D = 'd',
  E = 'e',
  F = 'f',
  G = 'g',
  H = 'h',
  I = 'i',
  J = 'j',
  K = 'k',
  L = 'l',
  M = 'm',
  N = 'n',
  O = 'o',
  P = 'p',
  Q = 'q',
  R = 'r',
  S = 's',
  T = 't',
  U = 'u',
  V = 'v',
  W = 'w',
  X = 'x',
  Y = 'y',
  Z = 'z',
  NUM_0 = '0',
  NUM_1 = '1',
  NUM_2 = '2',
  NUM_3 = '3',
  NUM_4 = '4',
  NUM_5 = '5',
  NUM_6 = '6',
  NUM_7 = '7',
  NUM_8 = '8',
  NUM_9 = '9',
  QUESTION = '?',
  CLICK = 'click',
}

export interface TriggerSpec {
  readonly type: TriggerType;
  readonly alt?: boolean|undefined;
  readonly ctrl?: boolean|undefined;
  readonly meta?: boolean|undefined;
  readonly shift?: boolean|undefined;
}

export function isKeyTrigger(triggerType: TriggerType): boolean {
  return triggerType !== TriggerType.CLICK;
}