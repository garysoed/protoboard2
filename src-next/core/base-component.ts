import {flattenResolver, ImmutableResolver, MutableResolver} from 'gs-tools/export/state';
import {instanceofType} from 'gs-types';
import {renderTheme} from 'mask';
import {Context, Ctrl, ivalue} from 'persona';
import {IValue, UnresolvedIO} from 'persona/export/internal';
import {EMPTY, merge, Observable, of, OperatorFunction, pipe} from 'rxjs';
import {switchMap, withLatestFrom} from 'rxjs/operators';

import {onTrigger} from '../trigger/trigger';
import {ComponentState} from '../types/component-state';
import {TriggerSpec} from '../types/trigger-spec';


type ActionFn<S extends ComponentState> =
    (context: Context<BaseComponentSpecType<S>>) => OperatorFunction<unknown, unknown>;
// type ActionFactory<C extends TriggerConfig, O> = (params: ActionParams<C, O>) => Action;

export interface BaseComponentSpecType<S extends ComponentState> {
  host: {
    readonly state: UnresolvedIO<IValue<ImmutableResolver<S>|undefined, 'state'>>;
  };
}

export function create$baseComponent<S extends ComponentState>(): BaseComponentSpecType<S> {
  return {
    host: {
      state: ivalue('state', instanceofType<ImmutableResolver<S>>(Object)),
    },
  };
}


// const LOG = new Logger('pb.core.BaseComponent');

export abstract class BaseComponent<S extends ComponentState> implements Ctrl {
  constructor(
      private readonly $baseComponent: Context<BaseComponentSpecType<S>>,
  ) {
    // this.setupActions();
  }

  protected installAction(
      action: ActionFn<S>,
      target$: Observable<HTMLElement>,
      triggerSpec$: Observable<TriggerSpec>,
      onCall$: Observable<unknown>,
  ): Observable<unknown> {
    return merge(target$.pipe(onTrigger(triggerSpec$)), onCall$)
        .pipe(action(this.$baseComponent));
  }

  protected get state(): ImmutableResolver<S> {
    return flattenResolver(this.$baseComponent.host.state);
  }


  // /**
  //  * Emits the current object ID of the host element, if any. If not, this doesn't emit any.
  //  */
  // @cache()
  // get objectPath$(): Observable<ObjectPath<O>> {
  //   return this.objectPathInput.getValue(this.context).pipe(
  //       switchMap(objectPath => {
  //         if (!objectPath) {
  //           LOG.warning('No object-path found');
  //           return EMPTY;
  //         }

  //         return of(objectPath);
  //       }),
  //   );
  // }

  // @cache()
  // get objectSpec$(): ImmutableResolver<O> {
  //   return $stateService.get(this.vine)._(this.objectPath$);
  // }

  get runs(): ReadonlyArray<Observable<unknown>> {
    return [renderTheme(this.$baseComponent)];
  }

  updateState<T>(
      mapFn: (resolverFrom: ImmutableResolver<S>) => MutableResolver<T>,
  ): OperatorFunction<T, unknown> {
    return pipe(
        withLatestFrom(this.$baseComponent.host.state),
        switchMap(([value, mutable]) => {
          if (!mutable) {
            return EMPTY;
          }

          return of(value).pipe(mapFn(mutable).set());
        }),
    );
  }

  // private setupActions(): void {
  //   this.addSetup(defer(() => {
  //     const obs: Array<Observable<unknown>> = [];
  //     const actionDescriptions: Array<Observable<ActionTrigger>> = [];
  //     for (const actionSpec of this.actions as readonly ActionSpec[]) {
  //       actionDescriptions.push(
  //           actionSpec.triggerSpec$.pipe(
  //               map(trigger => ({actionName: actionSpec.actionName, trigger})),
  //           ),
  //       );
  //       obs.push(this.setupTrigger(actionSpec));
  //     }

  //     const actionTriggers$ = actionDescriptions.length <= 0 ? of([]) : combineLatest(actionDescriptions);
  //     obs.push(this.setupTrigger(this.createActionSpec(
  //         helpAction,
  //         actionTriggers$.pipe(
  //             map(actions => {
  //               const hostEl = this.context.shadowRoot.host;
  //               return {
  //                 helpContent: {
  //                   tag: hostEl.tagName,
  //                   actions,
  //                 },
  //                 targetEl: hostEl,
  //                 trigger: {type: TriggerType.QUESTION, shift: true},
  //               };
  //             }),
  //         ),
  //         'Help',
  //     )));

  //     return merge(...obs);
  //   }));
  // }

  // private setupTrigger(
  //     actionSpec: ActionSpec,
  // ): Observable<unknown> {
  //   return actionSpec.trigger$.pipe(
  //       actionSpec.action,
  //   );
  // }
}
