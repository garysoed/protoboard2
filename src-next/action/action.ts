import {Context} from 'persona';
import {OperatorFunction} from 'rxjs';

import {BaseComponentSpecType} from '../core/base-component';
import {ComponentState} from '../types/component-state';


export type Action<S extends ComponentState, C> = (
    context: Context<BaseComponentSpecType<S>>,
    config: C,
) => OperatorFunction<unknown, unknown>;