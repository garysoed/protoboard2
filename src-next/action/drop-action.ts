import {filterNonNullable} from 'gs-tools/export/rxjs';
import {flattenResolver} from 'gs-tools/export/state';
import {Context} from 'persona';
import {concat, of, OperatorFunction, pipe} from 'rxjs';
import {map, switchMap, withLatestFrom} from 'rxjs/operators';

import {$activeState} from '../core/active-spec';
import {BaseComponentSpecType} from '../core/base-component';
import {RegionState} from '../types/region-state';


export function dropAction($: Context<BaseComponentSpecType<RegionState>>): OperatorFunction<unknown, unknown> {
  const activeState = $activeState.get($.vine);
  const activeContentIds = activeState.$('contentIds');
  const regionContentIds = flattenResolver($.host.state).$('contentIds');
  return pipe(
      withLatestFrom(regionContentIds, activeContentIds),
      map(([, regionIds, activeIds]) => {
        const newActiveIds = [...activeIds];
        const [droppedId] = newActiveIds.splice(0, 1);
        if (droppedId === undefined) {
          return null;
        }

        return {
          regionContents: [...regionIds, droppedId],
          activeContents: newActiveIds,
        };
      }),
      filterNonNullable(),
      switchMap(({activeContents, regionContents}) => {
        return concat(
            of(activeContents).pipe(activeContentIds.set()),
            of(regionContents).pipe(regionContentIds.set()),
        );
      }),
  );
}
