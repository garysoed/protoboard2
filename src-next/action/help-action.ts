import {Context, ievent} from 'persona';
import {Observable, OperatorFunction, pipe} from 'rxjs';
import {switchMap, tap, withLatestFrom} from 'rxjs/operators';

import {BaseComponentSpecType} from '../core/base-component';
import {TriggerEvent} from '../trigger/trigger-event';

import {HelpContent, ShowHelpEvent, SHOW_HELP_EVENT} from './show-help-event';


export interface HelpActionConfig {
  readonly helpContent: HelpContent;
}

export function helpAction(
    $: Context<BaseComponentSpecType<unknown>>,
    config$: Observable<HelpActionConfig>,
): OperatorFunction<TriggerEvent, unknown> {
  return pipe(
      withLatestFrom(config$),
      tap(([, config]) => {
        $.element.dispatchEvent(new ShowHelpEvent([config.helpContent]));
      }),
  );
}


export function forwardHelpEvent(
    config$: Observable<HelpActionConfig>,
): OperatorFunction<HTMLElement, unknown> {
  return pipe(
      switchMap(element => ievent(SHOW_HELP_EVENT, ShowHelpEvent, {matchTarget: false})
          .resolve(element)
          .value$),
      withLatestFrom(config$),
      tap(([event, config]) => {
        event.addContent(config.helpContent);
      }),
  );
}