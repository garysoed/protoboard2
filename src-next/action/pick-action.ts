import {flattenResolver} from 'gs-tools/export/state';
import {Context} from 'persona';
import {OperatorFunction, pipe} from 'rxjs';
import {map, withLatestFrom} from 'rxjs/operators';

import {$activeState} from '../core/active-spec';
import {BaseComponentSpecType} from '../core/base-component';
import {ComponentState} from '../types/component-state';


export function pickAction($: Context<BaseComponentSpecType<ComponentState>>): OperatorFunction<unknown, unknown> {
  const activeState = $activeState.get($.vine);
  return pipe(
      withLatestFrom(flattenResolver($.host.state)._('id'), activeState.$('contentIds')),
      map(([, id, contentIds]) => {
        return [...contentIds, id];
      }),
      activeState.$('contentIds').set(),
  );
}
