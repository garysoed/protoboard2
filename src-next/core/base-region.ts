import {filterNonNullable} from 'gs-tools/export/rxjs';
import {ImmutableResolver, MutableResolver} from 'gs-tools/export/state';
import {undefinedType} from 'gs-types';
import {Context, icall, ievent, ivalue, RenderSpec} from 'persona';
import {IAttr, ICall, IValue, OEvent, UnresolvedIO} from 'persona/export/internal';
import {Observable, OperatorFunction} from 'rxjs';
import {filter, map, switchMap, withLatestFrom} from 'rxjs/operators';

import {ActionEvent, ACTION_EVENT} from '../action/action-event';
import {dropAction} from '../action/drop-action';
import {pickAction} from '../action/pick-action';
import {renderContents} from '../render/render-contents';
import {RegionState} from '../types/region-state';
import {TriggerSpec, TriggerType, TRIGGER_SPEC_TYPE} from '../types/trigger-spec';

import {BaseComponent, BaseComponentSpecType, create$baseComponent} from './base-component';


interface BaseRegionSpecType<S extends RegionState> extends BaseComponentSpecType<S> {
  host: {
    readonly label: UnresolvedIO<IAttr>;
    readonly onAction: UnresolvedIO<OEvent<ActionEvent>>;
    readonly state: UnresolvedIO<IValue<ImmutableResolver<S>|undefined, 'state'>>;
    readonly drop: UnresolvedIO<ICall<undefined, 'drop'>>;
    readonly dropConfig: UnresolvedIO<IValue<TriggerSpec, 'dropConfig'>>;
  }
}

export function create$baseRegion<S extends RegionState>(): BaseRegionSpecType<S> {
  return {
    host: {
      ...create$baseComponent<S>().host,
      drop: icall('drop', undefinedType),
      dropConfig: ivalue('dropConfig', TRIGGER_SPEC_TYPE, {type: TriggerType.D}),
    },
  };
}

export abstract class BaseRegion<S extends RegionState> extends BaseComponent<S> {
  constructor(
      private readonly $baseRegion: Context<BaseRegionSpecType<S>>,
      defaultComponentName: string,
  ) {
    super($baseRegion, defaultComponentName);
  }

  get runs(): ReadonlyArray<Observable<unknown>> {
    return [
      ...super.runs,
      this.setupRenderContents(),
      this.setupHandlePick(),
      this.installAction(
          dropAction,
          'Drop',
          this.target$,
          this.$baseRegion.host.dropConfig,
          this.$baseRegion.host.drop,
      ),
    ];
  }

  abstract renderContents(): OperatorFunction<readonly RenderSpec[], unknown>;

  protected abstract get target$(): Observable<HTMLElement>;

  private setupHandlePick(): Observable<unknown> {
    const contentIds = this.state.$('contentIds') as unknown as MutableResolver<ReadonlyArray<{}>>;
    return this.target$.pipe(
        switchMap(target => ievent(ACTION_EVENT, ActionEvent).resolve(target).value$),
        filter(event => event.action === pickAction),
        withLatestFrom(contentIds),
        map(([event, contentIds]) => {
          if (contentIds.indexOf(event.id) < 0) {
            return null;
          }

          return contentIds.filter(content => content !== event.id);
        }),
        filterNonNullable(),
        contentIds.set(),
    );
  }

  private setupRenderContents(): Observable<unknown> {
    return (this.state as ImmutableResolver<RegionState>).$('contentIds').pipe(
        renderContents(this.$baseRegion.vine),
        this.renderContents(),
    );
  }
}