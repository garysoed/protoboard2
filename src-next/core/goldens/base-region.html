<pbt-test><gst-fake-shadow><div id="container"><!-- #ref -->onetwothree</div><link rel="stylesheet" href="base/test.css"></gst-fake-shadow></pbt-test>